rfcs/taskforce/0002-network-utility.md

```diff
@@ -0,0 +1,108 @@
+
+### Keypair Management
+
+For new keypair generation, we can use the new client-sdk which is a much lighter dependency and has a clean JavaScript interface.
+
+```
+type keypair = {
+  publicKey: string,
+  privateKey: option(string)
+}
+```
+
+However there are also existing keys which we'll need to be able to reference by public key alone since we don't control the private keys. These include community owned wallets for inclusion in the genesis ledger.The basic type of a keypair consists of the following:
+
+### Keyset Management
+
+Keysets are simply a collection of keypairs. We use these for categorizing different groups of keys that will be used in the lifecycle of a network. For each keypair in a keyset, we MAY or MAY NOT have the corresponding private keys. For each keyset, we may want to allocate a different amount of funds to their corresponding accounts in a genesis ledger.
+
+```
+type keysetMember = | Owned(keypair) | Unowned(publicKey)
+```
+
+Where it's implied that `type publicKey = string`.
+
+### Runtime Ledger Generation
+
+The code for generating the Runtime Genesis Ledger should be ported over from the [existing code](https://github.com/CodaProtocol/coda-automation/blob/master/scripts/testnet-keys.py#L203) but with the added flexibility of being able to inject any variety of keysets into the ledger with corresponding initial amounts.
+
+### Configuration Deployment
+
+The configuration deployment should be updated to take advantage of the kubernetes JavaScript client API. This tool should expose similar functionality to the existing tool, allowing users to upload keypairs to the kubernetes ConfigMap however not splitting out the task into predefined groups but rather allowing arbitrary keysets to be deployed.
+
+### Network Deployment and Monitoring
+
+This tool could optionally wrap the coda-automation tools and provide a simple interface for deploying a full testnet from a single command.
+
+## External Users
+
+Ideally this tool should be an easy starting point for users that are trying to run large node operations and particularly staking operations who wish to deploy complex infrastructure.
+
+## Explicitly NOT in scope for this project
+
+- Multi-cloud support
+- Local machine multi-node testnets
+- Moving away from kubernetes for configuration distribution
+
+## Alternatives Considered
+
+**Clean up existing python script**
+
+Pros:
+
+- Less work
+- Most existing network automation is already written in python
+
+Cons:
+
+- Requires lots of rewriting from the ground up in an untyped language
+- Less clean integrating with the client-sdk
+
+# Outstanding Questions
+
+- Are there any more usecases this tool should cover or have we over extended the scope of this project.
+
+- Carey has played with a couple frameworks for writing CLI tools but they all seem over complicated and unneccissary, should they still be considered?
```

Comment:
We should probably use some library for dealing with commandline parsing, flags, etc. Maybe there is a popular one that doesn't suck that already has reasonml bindings? Though, in a quick google search I didn't find one.

Reply:
Which frameworks were tried ? I can survey other options.

Reply:
I would highly recommend looking more at Cmdliner. I have had a lot of experience with Cmdliner and have found it to be simple to use and it gives you a lot right out of the box. You can also easily process the Cmdliner configurations you build up with custom code to do some neat programmatic cli stuff (such as cli fuzzing). The one downside of it is that the documentation is poor -- it's not that there isn't any, rather that it isn't very well written (at least as of last time I used it). But I would be willing to assist with getting it setup initially, and once you get the basics down, it is rather simple while still very flexible.

Reply:
I was going to say that this probably wouldn't work with bucklescript, but there is a library that does support it! https://github.com/ELLIOTTCABLE/bs-cmdliner

Reply:
Yeah @Schmavery recommended that too, that's my leading option now. Will push updates with this feedback and more details.